### PR TITLE
command/state mv

### PR DESCRIPTION
**Pull request by mitchellh**
This introduces the **state mv** command, a state plumbing subcommand for moving items within a state. This command can also move items from one state to a completely different state file. 

The only difference from `mv` (POSIX) is that we disallow _overwriting_ items in the state. We can choose to relax this at some point but in the name of safety we recommend doing a `terraform state rm` first if you want to overwrite something. I felt the behavior of simply "losing" the overwritten item was a bad default. 

I'm not particularly proud of the implementation but I believe it is well tested and it is easy to add more test cases for any edges we find. If we find a more elegant way to implement this in the future, we can refactor with confidence due to the tests covering both core and the CLI. 

The website examples are pasted below.

## Example: Rename a Resource

The example below renames a single resource:

```
$ terraform state mv aws_instance.foo aws_instance.bar
```

## Example: Move a Resource Into a Module

The example below moves a resource into a module. The module will be
created if it doesn't exist.

```
$ terraform state mv aws_instance.foo module.web
```

## Example: Move a Module Into a Module

The example below moves a module into another module.

```
$ terraform state mv module.foo module.parent.module.foo
```

## Example: Move a Module to Another State

The example below moves a module into another state file. This removes
the module from the original state file and adds it to the destination.
The source and destination are the same meaning we're keeping the same name.

```
$ terraform state mv -state-out=other.tfstate \
    module.web module.web
```

### CODE REVIEW

**Comment by mitchellh** [comment]
There is a single failing test in `command/` but it appears to also exist on `dev-0.7`. So the tests are good.

**Comment by marceldegraaf** [comment]
This is great! Thanks @mitchellh :tada:

File: command/state_mv.go [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+
+`
+	return strings.TrimSpace(helpText)
+}
+
+func (c *StateMvCommand) Synopsis() string {
+	return "Move an item in the state"
+}
+
+const errStateMv = `Error moving state: %[1]s
```
Comment (phinze) [inline command/state_mv.go]: Fancy fmt syntax or typo? 😁

File: terraform/state.go [hunk 2]
```diff
@@ -677,11 +673,13 @@ func (m *ModuleState) deepcopy() *ModuleState {
...
-		Path:      make([]string, len(m.Path)),
-		Outputs:   make(map[string]interface{}, len(m.Outputs)),
-		Resources: make(map[string]*ResourceState, len(m.Resources)),
+		Path:         make([]string, len(m.Path)),
+		Outputs:      make(map[string]interface{}, len(m.Outputs)),
+		Resources:    make(map[string]*ResourceState, len(m.Resources)),
+		Dependencies: make([]string, len(m.Dependencies)),
 	}
 	copy(n.Path, m.Path)
+	copy(n.Dependencies, m.Dependencies)
```
Comment (phinze) [inline terraform/state.go]: This change is an interesting one. Is this a bugfix or just a new need?

File: command/state_mv.go (hunk 1 above)
Comment (mitchellh) [inline command/state_mv.go]: Fany fmt syntax! it is positional references! (1-indexed).

File: terraform/state.go (hunk 2 above)
Comment (mitchellh) [inline terraform/state.go]: Yep actually just a bug that we never copied those over in the copy function and we needed them for the mv.

**Pull request closed without merging**